RearrangingCars.java

```diff
@@ -55,6 +55,68 @@ public int hashCode() {
 }
 
 public class RearrangingCars {
+    public static ArrayList<Move> optimisedRearrange(int[] originalOrder, int[] finalOrder) {
+        if (originalOrder == null || finalOrder == null || originalOrder.length != finalOrder.length)
+            return null;
+
+        // Idea: Repeatedly replace current empty spot with the car that has to go there
+
+        ArrayList<Move> moves = new ArrayList<>();
+        if (originalOrder.length == 0) return moves;
+
+        HashMap<Integer, Integer> positions = new HashMap<>();
+        int N = originalOrder.length;
+        int[] originalOrderCopy = originalOrder.clone();
+        for (int i = 0; i < N; i++) {
+            positions.put(originalOrderCopy[i], i);
+        }
+
```

Comment:
You could consider breaking the code down to some parts. That way testing each one would be much easier. 
Either way it's fine I think, this looks very nice too. On the tests consider adding a specific example, and test what you expect from the new rearranging to happen, just like you had done initially for testing the first solution :)